tools/flash_shelly.py

```diff
@@ -0,0 +1,501 @@
+PURPLE = '\033[1;95m' if not arch.lower().startswith('win') else ""
+NC = '\033[0m' if not arch.lower().startswith('win') else ""
+
+if not importlib.util.find_spec("zeroconf"):
+  logger.info('Installing zeroconf...')
+  pipe = subprocess.check_output(['pip3', 'install', 'zeroconf'])
+  import zeroconf
+else:
+  import zeroconf
+
+if not importlib.util.find_spec("requests"):
+  logger.info('Installing requests...')
+  pipe = subprocess.check_output(['pip3', 'install', 'requests'])
+  import requests
+else:
+  import requests
+
+class MyListener:
+  def __init__(self):
+    self.device_list = []
+    self.p_list = []
+
+  def add_service(self, zeroconf, type, name):
+      self.device_list.append(name.replace('._http._tcp.local.', ''))
+      # info = zeroconf.get_service_info(type, name, 2000)
+      # logger.trace('INFO: %s' % info)
+      # properties = { y.decode('ascii'): info.properties.get(y).decode('ascii') for y in info.properties.keys() }
+      # self.p_list.append(properties)
+      # logger.trace('properties: %s' % properties)
+      # json_object = json.dumps(properties, indent = 2)
+
+def shelly_model(type, mode):
+  if mode != 'stock':
+    options = {'SHSW-1' : 'Shelly1',
+               'switch1' : 'Shelly1',
+               'SHSW-PM' : 'Shelly1PM',
+               'switch1pm' : 'Shelly1PM',
+               'SHSW-21' : 'Shelly2',
+               'switch2' : 'Shelly2',
+               'SHSW-25' : 'Shelly25',
+               'switch25' : 'Shelly25',
+               'SHPLG-S' : 'ShellyPlugS',
+               'shelly-plug-s' : 'ShellyPlugS',
+               'SHDM-1' : 'ShellyDimmer',
+               'dimmer1' : 'ShellyDimmer',
+               'SHRGBW2' : 'ShellyRGBW2',
+               'rgbw2' : 'ShellyRGBW2',
+    }
+  else:
+    options = {'switch1' : 'SHSW-1',
+               'switch1pm' : 'SHSW-PM',
+               'switch2' : 'SHSW-21',
+               'switch25' : 'SHSW-25',
+               'shelly-plug-s' : 'SHPLG-S',
+               'dimmer1' : 'SHDM-1',
+               'rgbw2' : 'SHRGBW2',
+    }
+  return(options[type])
+
+
+def parseVersion(vs):
+  pp = vs.split("-");
+  v = pp[0].split(".");
+  variant = ""
+  varSeq = 0
+  if len(pp) > 1:
+    i = 0
+    for x in pp[1]:
+      c = pp[1][i]
+      if not c.isnumeric():
+        variant += c
+      else:
+        break
+      i += 1
+    varSeq = int(pp[1][i]) or 0
+  major = int(v[0])
+  minor = int(v[1])
+  patch = int(v[2])
+  return (major, minor, patch, variant, varSeq)
+
+def isNewer(v1, v2):
+  vi1 = parseVersion(v1)
+  vi2 = parseVersion(v2)
+  if (vi1[0] != vi2[0]):
+    return (vi1[0] > vi2[0])
+  elif (vi1[1] != vi2[1]):
+    return (vi1[1] > vi2[1])
+  elif (vi1[2] != vi2[2]):
+    return (vi1[2] > vi2[2])
+  elif (vi1[3] != vi2[3]):
+    return True
+  elif (vi1[4] != vi2[4]):
+    return (vi1[4] > vi2[4])
+  else:
+    return False
+
+
+def write_flash(device, lfw, dlurl, cfw_type, mode):
+  logger.debug("\n" + WHITE + "write_flash" + NC)
```

Comment:
using + to build strings is not pythonic.
since this is python 3, we can use the `f"..."` operator taht will perform variable interpolation, e.g.

logger.debug(f"\n{WHITE}write_flash{NC}")

much nicer and more readable

Reply:
done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/ed7b9f018b7e77c98a3700db2e41a9753a03992d